### PR TITLE
[Fix] Improve merge for recognition of long words (#1936)

### PR DESCRIPTION
**Pull request by Razlaw**
Hello again @felixdittrich92 :)

I took a deeper look into the splitting and merging of the crops for the recognition and came up with a first version we can have a look at.
I tried to stick to the original approach and keep it simple.
You've had the idea to dynamically compute the hyperparameters for the aspect ratios per crop.
That's not implemented yet, but we I guess this PR is work in progress anyway :)

The main idea so far is:
- Split the crop into equally sized sub-crops with known overlap ratios
- Use the known overlap of neighboring sub-crops to estimate the overlap when merging recognized strings

This worked well for the examples I provided in the issue #1936.

Sub-crops right before being processed by recognizer - with recognized string per sub-crop:
crops_for_recognition_after_preprocessing_batch_0_file_0
2025-02-20T
crops_for_recognition_after_preprocessing_batch_0_file_1
2-20T06:51:
crops_for_recognition_after_preprocessing_batch_0_file_2
06:51:13.000
crops_for_recognition_after_preprocessing_batch_0_file_3
:51:13.000Z
Merged result: 2025-02-20T06:51:13.000Z 
:white_check_mark: 

crops_for_recognition_after_preprocessing_batch_0_file_0
XUrA+b3iHQP
crops_for_recognition_after_preprocessing_batch_0_file_1
3iHQP+F3qxu
crops_for_recognition_after_preprocessing_batch_0_file_2
+F3qxuOZZDR:
crops_for_recognition_after_preprocessing_batch_0_file_3
OZZDRA/628+
crops_for_recognition_after_preprocessing_batch_0_file_4
RA/6Z8+6nbf
Merged result: XUrA+b3iHQP+F3qxuOZZDRA/628+6nbf 
Not perfectly correct but at least correctly merged :)

I used the evaluation script from the repo to check how the changes affect the scores - thanks for providing that script btw.
To summarize, the results are actually not that different to the existing approach..
But at least significantly faster on the CORD dataset.
I've looked into the data and my hypothesis is that there are mostly short words and numbers that do not benefit from the split and merge.

Additionally I ran an evaluation on a dataset we labeled ourselves.
It consists of 160 photographies of german shopping receipts - so it includes at least some lengthy words ;)
The average accuracy improved by 1%.
For dates - consisting of the long datetimes as shown in the example, but also shorter versions of just time or date - the accuracy improved by ~7%.
It's not much, but at least something :)

I'm eager to hear your feedback about the changes. No pressure though ;) 

---

Appendix:
The full results of the evaluation on FUNSD and CORD
Variant 2 and 3 use the same critical_ar and target_ar parameters of 8 and 6 - the default from docTR - to generate sub-crops of comparable sizes. Mostly relevant to compare the run durations. 
Variant 4 uses critical_ar of 4 and target_ar of 4, because this utilizes the input window (32 times 128 pixels) of the recognizer fully.
This variant worked best on our data, but had slightly worse results on FUNSD and CORD.

- Models: db_resnet50 + parseq
- **FUNSD**
	- **1 docTR main**
		- Text Recognition - Accuracy: 86.67% (unicase: 87.47%) OCR - Recall: 73.46% (unicase: 73.99%), Precision: 76.20% (unicase: 76.75%), Mean IoU: 68.00%  
			- 149/149 [30:44<00:00, 12.38s/it]
			- 50/50 [12:03<00:00, 14.48s/it]
	- **2 docTR main - only with fixed computation of number of sub-crops**
		- Text Recognition - Accuracy: 86.46% (unicase: 87.25%) OCR - Recall: 73.34% (unicase: 73.87%), Precision: 76.08% (unicase: 76.63%), Mean IoU: 68.00%  
			- 149/149 [31:42<00:00, 12.77s/it]
			- 50/50 [12:17<00:00, 14.75s/it]
	- **3 Fix with crop parameters similar to version 2**
		- Text Recognition - Accuracy: 86.60% (unicase: 87.40%) OCR - Recall: 73.45% (unicase: 73.98%), Precision: 76.19% (unicase: 76.74%), Mean IoU: 68.00%  
			- 149/149 [32:01<00:00, 12.90s/it]
			- 50/50 [12:14<00:00, 14.69s/it]
	- **4 Fix with full utilization of recognition input window**
		- Text Recognition - Accuracy: 85.71% (unicase: 86.51%) OCR - Recall: 73.14% (unicase: 73.63%), Precision: 75.87% (unicase: 76.38%), Mean IoU: 68.00%

- **CORD - ran on different machine so timings not comparable to FUNSD**
	- **1 docTR main**
		- Text Recognition - Accuracy: 91.71% (unicase: 92.07%) OCR - Recall: 85.03% (unicase: 85.34%), Precision: 79.30% (unicase: 79.58%), Mean IoU: 73.00%
			- 800/800 [38:29<00:00,  2.89s/it]
			- 100/100 [04:40<00:00,  2.80s/it]
	- **2 docTR main - only with fixed computation of number of sub-crops**
		- Text Recognition - Accuracy: 91.68% (unicase: 92.04%) OCR - Recall: 85.03% (unicase: 85.32%), Precision: 79.30% (unicase: 79.57%), Mean IoU: 73.00%
			- 800/800 [38:22<00:00,  2.88s/it]
			- 100/100 [04:38<00:00,  2.78s/it]
	- **3 Fix with crop parameters similar to version 2**
		- Text Recognition - Accuracy: 91.72% (unicase: 92.08%) OCR - Recall: 85.09% (unicase: 85.39%), Precision: 79.35% (unicase: 79.63%), Mean IoU: 73.00%  
			- 800/800 [27:01<00:00,  2.03s/it]
			- 100/100 [03:19<00:00,  1.99s/it]
		- Significantly faster with similar results - maybe because of Hamming distance instead of Levenshtein
	- **4 Fix with full utilization of recognition input window**
		- Text Recognition - Accuracy: 90.16% (unicase: 90.57%) OCR - Recall: 84.25% (unicase: 84.55%), Precision: 78.57% (unicase: 78.85%), Mean IoU: 73.00%
			- 800/800 [38:36<00:00,  2.90s/it]
			- 100/100 [04:30<00:00,  2.71s/it]

### CODE REVIEW

**Comment by felixdittrich92** [comment]
Hey @Razlaw :wave:,

wow thanks a lot for the PR and the detailed benchmarks I will have a look on monday :+1: Have a nice weekend :)

**Comment by Razlaw** [comment]
Nice job on the refactorings :) 
I've put them in some additional commits, so we can talk about the new state in more detail.
In the end, I'll clean and squash them.

Thanks for the test files :+1: 
Do you have any ground truth labels for those files? Then I wouldn't need to check "by hand" where the merging failed ^^

File: doctr/models/recognition/utils.py
[the remark this answers is not in the document]
Comment (Razlaw) [reply doctr/models/recognition/utils.py:1]: > multi_string: (["aaaa", "aaab", "aabc"], 0.8, 0.3, "aaaabc"), <- IndexError

Interesting, I didn't get an IndexError for the first example, but there was a mistake for the repetitions. That mistake should be fixed now.

> string_merge: ("a", "b", 0.5, "ab"), <-- empty

A added a quick fix for the second example - it's very much up for discussion though. Anyhow, that example is probably an edge case when it comes to splitting long words and merging them back together ;)

File: tests/common/test_models_recognition_utils.py [hunk 1]
```diff
@@ -1,31 +1,206 @@
 import pytest
 
-from doctr.models.recognition.utils import merge_multi_strings, merge_strings
+from doctr.models.recognition.utils import _get_index_closest_to_target_index, merge_multi_strings, merge_strings
 
 
+# Last character of first string and first of last string will be cropped when merging - indicated by X
 @pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (Razlaw) [reply tests/common/test_models_recognition_utils.py:7]: I'm not really sure if I understand your comment correctly.
The overlap ratio is what's being used to estimate where to merge the strings.
So we need to adapt it since the strings in the test cases change, right?

File: tests/common/test_models_recognition_utils.py [hunk 2]
```diff
@@ -1,31 +1,206 @@
...
+        (["abcdX", "XdefX", "XefghiX", "Xaijk"], 0.5, 0.8, "abcdefghijk"),
     ],
 )
-def test_merge_multi_strings(seq_list, merged):
-    assert merged == merge_multi_strings(seq_list, 1.4)
+def test_merge_multi_strings(seq_list, overlap_ratio, last_overlap_ratio, merged):
+    assert merged == merge_multi_strings(seq_list, overlap_ratio, last_overlap_ratio)
+
+
+def test_example_from_docs_1():
```
[the remark this answers is not in the document]
Comment (Razlaw) [reply tests/common/test_models_recognition_utils.py:37]: Yes, I will refactor the tests once the code is stable.

File: tests/common/test_models_recognition_predictor.py [hunk 3]
```diff
@@ -5,35 +5,160 @@
...
-    preds = remap_preds(preds, crop_map, dilation)
+def test_remap_preds(preds, crop_map, split_overlap_ratio, pred):
+    preds = remap_preds(preds, crop_map, split_overlap_ratio)
     assert len(preds) == len(pred)
     assert preds == pred
     assert all(isinstance(pred, tuple) for pred in preds)
     assert all(isinstance(pred[0], str) and isinstance(pred[1], float) for pred in preds)
+
+
+def test_dont_split_if_not_necessary():
```
[the remark this answers is not in the document]
Comment (Razlaw) [reply tests/common/test_models_recognition_predictor.py:42]: Same here, I would do that once the code is in a stable state.

> Additional it should work with both channels_last = True | False the same way

Just to make sure I understand correctly. There is one test case that has the channels_last parameter set to False. Would you like to have all other test cases in a version with True and False?

File: tests/common/test_models_recognition_utils.py [hunk 4]
```diff
@@ -136,12 +136,12 @@ def test_multiple_repetitions_shorter():
 
 
 def test_multiple_repetitions_full_overlap():
-    a_ = "ababX"
-    b = "Xabab"
```
Comment (Razlaw) [inline tests/common/test_models_recognition_utils.py]: These were impossible, because b is supposed to be merged to a from the right, but contains a character more on its left side than a does.

File: doctr/models/recognition/predictor/_utils.py [hunk 5]
```diff
@@ -17,104 +17,126 @@ def split_crops(
...
+        A tuple containing:
+            - The new list of crops (possibly with splits),
+            - A mapping indicating how to reassemble predictions,
+            - A boolean indicating whether remapping is required.
     """
-    _remap_required = False
-    splits_map: list[int | tuple[int, int, float]] = []
+    remap_required = False
     new_crops: list[np.ndarray] = []
+    crop_map: list[int | tuple[int, int]] = []
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/models/recognition/predictor/_utils.py]: Correct good catch :+1: 
Yeah I hacked this together in my lunch break .. that you don't have to wait to long for the review 😂

File: tests/common/test_models_recognition_predictor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply tests/common/test_models_recognition_predictor.py:42]: Keep up with one test for `channels_last=False` :+1:

I faiced an incositents which goes trough the whole docTR lib .... 

We shouldn't except Tensors for the modular predictor's (incl. ocr_predictor) without handling them correctly (at least for pytorch) ... and maybe it would be also good to go with one fixed shape format (H, W, C) for the inference pipes 🙄
But that's out of scope for your PR

File: tests/common/test_models_recognition_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply tests/common/test_models_recognition_utils.py:7]: Forget this comment :sweat_smile: Everything right here :+1:

File: tests/common/test_models_recognition_utils.py [hunk 6]
```diff
@@ -11,7 +11,7 @@
         ("abcdX", "Xdef", 0.75, "abcdef"),
         ("abcdeX", "Xdef", 0.9, "abcdef"),
         ("abcdefX", "Xdef", 0.9, "abcdef"),
-        ("abccccX", "Xcccccc", 0.4, "abcccccccc"),
+        ("abccccX", "Xcccccc", 0.5, "abcccccccc"),
```
Comment (felixdittrich92) [inline tests/common/test_models_recognition_utils.py:14]: Anyway this test case is really hard to understand :sweat_smile: 

So later on each should get and comment about the merge area :)

File: doctr/models/recognition/utils.py [hunk 7]
```diff
@@ -59,7 +60,8 @@ def merge_strings(a: str, b: str, overlap_ratio: float) -> str:
         return a_crop + b_crop
 
     # Find best overlap by minimizing Hamming distance + distance from expected overlap size
-    combined_scores = [score + abs(i - expected_overlap) for i, score in enumerate(scores)]
+    expected_overlap_index = expected_overlap - 1
```
Comment (felixdittrich92) [inline doctr/models/recognition/utils.py:65]: I think it would be better to adjust directly here (L43):

`expected_overlap = round(len(b) * overlap_ratio) - 2` to -3 ?

**Comment by felixdittrich92** [comment]
> Nice job on the refactorings :) I've put them in some additional commits, so we can talk about the new state in more detail. In the end, I'll clean and squash them.
> 
> Thanks for the test files 👍 Do you have any ground truth labels for those files? Then I wouldn't need to check "by hand" where the merging failed ^^

Hey @Razlaw :)

Yep sounds good to me :+1: 
I double checked and it doesn't fails on split ..the model produces garbage in the most cases because of missing padding

Let me try to create annotations for the files today

File: tests/common/test_models_recognition_utils.py (hunk 6 above)
Comment (Razlaw) [reply tests/common/test_models_recognition_utils.py]: Yes, true, it was even confusing when I wrote the test ^^ 
I added a comment. Is that okay or should I add more explanations?

**Comment by Razlaw** [comment]
I didn't address the codecov warnings yet. There is one I'm not sure we can even address using a test. I'll leave another comment there.
Btw, you mentioned you wanted to test some possible further improvements. Would you like to keep the PR open?
Did you get the chance to generate the ground truth for the data you provided?

File: doctr/models/recognition/predictor/_utils.py [hunk 8]
```diff
@@ -15,69 +17,126 @@ def split_crops(
...
-            # Avoid sending zero-sized crops
-            _crops = [crop for crop in _crops if all(s > 0 for s in crop.shape)]
-            # Record the slice of crops
-            crop_map.append((len(new_crops), len(new_crops) + len(_crops)))
-            new_crops.extend(_crops)
-            # At least one crop will require merging
-            _remap_required = True
+                # Fallback: treat it as a single crop
+                crop_map.append(len(new_crops))
+                new_crops.append(crop)
```
Comment (Razlaw) [inline doctr/models/recognition/predictor/_utils.py:62]: Is it even possible to get here?
Addressing codecov warning "Added lines #L61 - L62 were not covered by tests"

Comment (felixdittrich92) [reply doctr/models/recognition/predictor/_utils.py:62]: Mh this is an real extrem edge case but could happen if `overlap_width>=split_width` 

So two options:

1. We add 1 test case which covers this 
2. We ignore covering this - because it'S an real extrem edge case

```
else:  # pragma: no cover
    # Fallback: treat it as a single crop
    crop_map.append(len(new_crops))
    new_crops.append(crop)
 ```

File: doctr/models/recognition/utils.py [hunk 9]
```diff
@@ -4,81 +4,90 @@
...
     """
-
-    def _recursive_merge(a: str, seq_list: list[str], dil_factor: float) -> str:
-        # Recursive version of compute_overlap
-        if len(seq_list) == 1:
-            return merge_strings(a, seq_list[0], dil_factor)
-        return _recursive_merge(merge_strings(a, seq_list[0], dil_factor), seq_list[1:], dil_factor)
-
-    return _recursive_merge("", seq_list, dil_factor)
+    if not seq_list:
```
Comment (felixdittrich92) [inline doctr/models/recognition/utils.py:86]: We can cover this by passing a empty list in the test cases :)

([], 0.5, 0.4, ""),

File: tests/common/test_models_recognition_predictor.py
Comment (felixdittrich92) [inline tests/common/test_models_recognition_predictor.py:1]: This looks really good :+1:

File: tests/common/test_models_recognition_utils.py [hunk 10]
```diff
@@ -4,28 +4,60 @@
...
     [
-        [["abc", "def"], "abcdef"],
-        [["abcd", "def", "efgh", "ijk"], "abcdefghijk"],
-        [["abcdi", "defk", "efghi", "aijk"], "abcdefghijk"],
+        # One character at each conjunction point will be cropped when merging - indicated by X
+        (["abcX", "Xdef"], 0.5, 0.5, "abcdef"),
+        (["abcdX", "XdefX", "XefghX", "Xijk"], 0.5, 0.5, "abcdefghijk"),
+        (["abcdX", "XdefX", "XefghiX", "Xaijk"], 0.5, 0.8, "abcdefghijk"),
+        (["aaaa", "aaab", "aabc"], 0.8, 0.3, "aaaabc"),
+        # Example from docs
```
Comment (felixdittrich92) [inline tests/common/test_models_recognition_utils.py:58]: same here :)

File: tests/common/test_models_recognition_utils.py [hunk 11]
```diff
@@ -4,28 +4,60 @@
...
+        ("abcdX", "Xdef", 0.75, "abcdef"),
+        ("abcdeX", "Xdef", 0.9, "abcdef"),
+        ("abcdefX", "Xdef", 0.9, "abcdef"),
+        # Long repetition - four of seven characters in the second string are in the estimated overlap
+        # X-chars will be cropped during merge, because they might be cut off during splitting of corresponding image
+        ("abccccX", "Xcccccc", 4 / 7, "abcccccccc"),
+        ("abc", "", 0.5, "abc"),
+        ("", "abc", 0.5, "abc"),
+        ("a", "b", 0.5, "ab"),
+        # Example from docs 1
```
Comment (felixdittrich92) [inline tests/common/test_models_recognition_utils.py:20]: What does this mean "Example from docs 1" ? :)

Let's refine the two comments "Example from docs 1" & "Example from docs 2" to something meaningful like the other comments :)

File: doctr/models/recognition/predictor/_utils.py [hunk 12]
```diff
@@ -15,69 +17,126 @@ def split_crops(
...
+        split_width: Width of each split.
+        overlap_width: Width of the overlapping region.
+        channels_last: Whether the image is in channels-last format.
+
+    Returns:
+        - A list of horizontal image slices.
+        - The actual overlap ratio of the last split.
+    """
+    image_width = image.shape[1] if channels_last else image.shape[-1]
+    if image_width <= split_width:
```
Comment (felixdittrich92) [inline doctr/models/recognition/predictor/_utils.py:87]: Let's add a tiny additional test for this :)

**Review by felixdittrich92 — CHANGES_REQUESTED** [review]
Hi @Razlaw :wave:,

Thanks a lot for the updates only a few minor things left and we are good to merge :+1:

File: tests/common/test_models_recognition_utils.py (hunk 11 above)
Comment (Razlaw) [reply tests/common/test_models_recognition_utils.py:20]: There is not much meaningful behind these test cases :)
They are just the usage examples from the docstrings of the methods
```
    """
    ...
    Example::
        >>> from doctr.models.recognition.utils import merge_strings
        >>> merge_strings('abcd', 'cdefgh', 0.5)
        'abcdefgh'
        >>> merge_strings('abcdi', 'cdefgh', 0.5)
        'abcdefgh'
    """
```
I wanted to make sure that those example cases work as intended.
I can change the comment at the test case to "Happy case" but that's not really meaningful either.
We can also remove those test cases if you want.
Any preference?

File: doctr/models/recognition/predictor/_utils.py (hunk 8 above)
Comment (felixdittrich92) [reply doctr/models/recognition/predictor/_utils.py:62]: Mh .. I think it would be better to clip the value in this case - to avoid an hard exception

Additional we could add a logger warning ..but I think this would spam a bit to much :sweat_smile:

File: tests/common/test_models_recognition_utils.py (hunk 11 above)
Comment (felixdittrich92) [reply tests/common/test_models_recognition_utils.py:20]: Let's drop :)

File: doctr/models/recognition/predictor/_utils.py (hunk 8 above)
Comment (Razlaw) [reply doctr/models/recognition/predictor/_utils.py:62]: Yep, logging sounds like spam :) 
On the other hand, I would want to notice if I set a parameter to a bad value -> pro exception or spam. 
On the other other hand, those parameters will not be tuned frequently and anyone playing around with them should roughly know what they do..
You decide ;)

File: doctr/models/recognition/utils.py (hunk 9 above)
Comment (Razlaw) [reply doctr/models/recognition/utils.py:86]: Done :+1:

File: doctr/models/recognition/predictor/_utils.py (hunk 8 above)
Comment (felixdittrich92) [reply doctr/models/recognition/predictor/_utils.py:62]: Yeah also true - then raise a ValueError because user input

**Comment by Razlaw** [comment]
Regarding the mypy errors: I guess they don't originate from the code that we changed in this PR, right? How should we proceed here?

**Comment by felixdittrich92** [comment]
> Regarding the mypy errors: I guess they don't originate from the code that we changed in this PR, right? How should we proceed here?

Correct - ignore them I will fix it later :)
